akka-projection-grpc/src/main/scala/akka/projection/grpc/consumer/scaladsl/EventProducerPushDestination.scala

```diff
@@ -0,0 +1,182 @@
+ * configured journal and can then be consumed through a local projection. A producer can push events for multiple
+ * entities but no two producer are allowed to push events for the same entity.
+ *
+ * The event consumer service is not needed for normal projections over gRPC where the consuming side can access and
+ * initiate connections to the producing side.
+ *
+ * Producers are started using the [[akka.projection.grpc.producer.scaladsl.EventProducerPush]] API.
+ */
+// FIXME Java API
+@ApiMayChange
+object EventProducerPushDestination {
+
+  /**
+   * @param acceptedStreamId Only accept this stream ids, deny others
+   */
+  def apply(acceptedStreamId: String): EventProducerPushDestination =
+    new EventProducerPushDestination(None, acceptedStreamId, (_, _) => Transformation.empty, None, immutable.Seq.empty)
+
+  @ApiMayChange
+  object Transformation {
+    val empty = new Transformation {
+      // Note: this is also the empty/identity transformation
+      override private[akka] def apply(eventEnvelope: EventEnvelope[Any]): EventEnvelope[Any] =
+        eventEnvelope
+    }
+  }
+
+  private final case class MapTags(f: EventEnvelope[Any] => Set[String]) extends Transformation {
+    override private[akka] def apply(eventEnvelope: EventEnvelope[Any]): EventEnvelope[Any] = {
+      val newTags = f(eventEnvelope)
+      if (newTags eq eventEnvelope.tags) eventEnvelope
+      else copyEnvelope(eventEnvelope)(tags = newTags)
+    }
+
+  }
+
+  private final case class MapPersistenceId(f: EventEnvelope[Any] => String) extends Transformation {
+    override private[akka] def apply(eventEnvelope: EventEnvelope[Any]): EventEnvelope[Any] = {
+      val newPid = f(eventEnvelope)
+      if (newPid eq eventEnvelope.persistenceId) eventEnvelope
+      else copyEnvelope(eventEnvelope)(persistenceId = newPid)
+    }
+  }
+
+  private final case class MapPayload(f: EventEnvelope[Any] => Option[Any]) extends Transformation {
+    override private[akka] def apply(eventEnvelope: EventEnvelope[Any]): EventEnvelope[Any] = {
+      val newMaybePayload = f(eventEnvelope)
+      if (newMaybePayload eq eventEnvelope.eventOption) eventEnvelope
+      else copyEnvelope(eventEnvelope)(eventOption = newMaybePayload)
+    }
+  }
+
+  private final case class AndThen(first: Transformation, next: Transformation) extends Transformation {
+    override private[akka] def apply(eventEnvelope: EventEnvelope[Any]): EventEnvelope[Any] =
+      next.apply(first.apply(eventEnvelope))
+  }
+
+  /**
+   * Transformation of events from the producer type to the internal representation stored in the journal
+   * and seen by local projections.
+   *
+   * Events can be excluded by mapping them to `None`.
+   */
+  sealed trait Transformation {
+    def registerTagMapper[Event](f: EventEnvelope[Event] => Set[String]): Transformation =
+      this.andThen(MapTags(f.asInstanceOf[EventEnvelope[Any] => Set[String]]))
```

Comment:
Regarding your question about the class tag. Isn't that needed here also, otherwise the user may register two mappers for `EventEnvelope[Foo]` and `EventEnvelope[Bar]`, and we will not see the difference and just pass in a Bar where it's expected a Foo. Nah, I think we need the same stuff.

Reply:
Alright, I'll add that then

Reply:
Aligned as much as I could in [39ef575](https://github.com/akka/akka-projection/pull/932/commits/39ef575d6895ecb009e79ac33328e24dc1d47003) but it can't be 1:1 because for example persistence id must be mapped also for filtered payloads to not get gaps, and then the payload type is not important. 

Also, because the transformations are more fine grained, envelope to pid, envelope to tags, it feels useful that they can be chained, rather than be just one per event class as in the other Transformation (which only maps payloads). So now both the typed and untyped transformations are applied if present, since you may map payload and then also pid but still need the mapped pid in the journal to avoid gaps.

Reply:
looking good